File: Hashmaps and Sets/Python/validAnagram.py
# Given two strings s and t return whether or not s is an anagram of t.
# Note: An anagram is a word formed by reordering the letters of another word.

# Ex: Given the following strings...

# s = "cat", t = "tac", return true
# s = "listen", t = "silent", return true
# s = "program", t = "function", return false

def validAnagram(s, t):
    # convert to dictionary with each character of s and t as keys and 
    # their count as values and check if both dictionaries are equal
    return dict((c, s.count(c)) for c in s) == dict((ch, t.count(ch)) for ch in t)


def test():
    testCases = [("cat", "tac"), ("listen", "silent"), ("program", "function")]
    expecteds = [True, True, False]
    for case, expected in zip(testCases, expecteds):
        actual = validAnagram(*case) # same as validAnagram(case[0], case[1])
        assert actual == expected, f"Wrong answer. Case: {case}. Expected: {expected}. Actual: {actual}."
    print("All test cases passed.")


test()

File: Hashmaps and Sets/JavaScript/validAnagram.js
// Given two strings s and t return whether or not s is an anagram of t.
// Note: An anagram is a word formed by reordering the letters of another word.

// Ex: Given the following strings...

// s = "cat", t = "tac", return true
// s = "listen", t = "silent", return true
// s = "program", t = "function", return false

const validAnagram = (s, t) => {
  let charMapS = {};
  s.split("").forEach((element) => {
    charMapS[element] = charMapS[element] ? charMapS[element] + 1 : 1;
  });
  let charMapT = {};
  t.split("").forEach((element) => {
    charMapT[element] = charMapT[element] ? charMapT[element] + 1 : 1;
  });
  return Object.entries(charMapS).sort().toString() === Object.entries(charMapT).sort().toString();
};


const test = () => {
    testCases = [
      { s: "cat", t: "tac" },
      { s: "silent", t: "listen" },
      { s: "program", t: "function" },
    ];
    expecteds = [true, true, false];
  
    for (let i = 0; i < testCases.length; i++) {
      const actual = validAnagram(testCases[i].s, testCases[i].t);
      console.log(`For '${testCases[i]}' output obtained '${actual}'`);
      console.assert(
        actual == expecteds[i],
        `Wrong answer. Expected: '${expecteds[i]}'. Found: '${actual}'`
      );
    }
    console.log("Test completed.");
  };
  
  test();
  